lib/api/multiObjectDelete.js

```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
     let numOfObjectsRemoved = 0;
     const skipError = new Error('skip');
     const objectLockedError = new Error('object locked');
+    let deleteFromStorage = [];
+    const initialStep = shouldBatchGetObjectsFromMetadata(bucketName, inPlay, log);
 
-    // doing 5 requests at a time. note that the data wrapper
-    // will do 5 parallel requests to data backend to delete parts
-    return async.forEachLimit(inPlay, 5, (entry, moveOn) => {
-        async.waterfall([
-            callback => {
-                let decodedVersionId;
-                if (entry.versionId) {
-                    decodedVersionId = entry.versionId === 'null' ?
-                        'null' : versionIdUtils.decode(entry.versionId);
-                }
-                if (decodedVersionId instanceof Error) {
-                    monitoring.promMetrics('DELETE', bucketName, 404,
-                        'multiObjectDelete');
-                    return callback(errors.NoSuchVersion);
-                }
-                return callback(null, decodedVersionId);
-            },
-            // for obj deletes, no need to check acl's at object level
-            // (authority is at the bucket level for obj deletes)
-            (versionId, callback) => metadataGetObject(bucketName, entry.key,
-                versionId, log, (err, objMD) => {
-                    // if general error from metadata return error
-                    if (err) {
-                        monitoring.promMetrics('DELETE', bucketName, err.code,
-                            'multiObjectDelete');
-                        return callback(err);
-                    }
-                    if (!objMD) {
-                        const verCfg = bucket.getVersioningConfiguration();
-                        // To adhere to AWS behavior, create a delete marker
-                        // if trying to delete an object that does not exist
-                        // when versioning has been configured
-                        if (verCfg && !entry.versionId) {
-                            log.debug('trying to delete specific version ' +
-                            ' that does not exist');
-                            return callback(null, objMD, versionId);
+    return async.waterfall([
+        callback => initialStep((err, cache) => callback(err, cache)),
+        (cache, callback) => async.forEachLimit(inPlay, 50, (entry, moveOn) => {
+            async.waterfall([
+                callback => callback(...processObjectVersion(entry, bucketName)),
+                // for obj deletes, no need to check acl's at object level
+                // (authority is at the bucket level for obj deletes)
+                (versionId, callback) => metadataGetObject(bucketName, entry.key,
```

Comment:
I would expect this `metadataGetObject` call not to happen when the entry is already in cache (due to `metadataGetObjects` called earlier to populate the cache), but I don't see the mechanism to check this and skip `metadataGetObject`.

Reply:
I wonder if this is worth changing the API architecture for this? Originally this was something we discussed with @francoisferrand specifically not to change the API structure while benefiting from the previous cache. I am fine with both approaches, but the current way seem less risky...

We could just wrap this call and do something like:

```js
const handleObjectMetadata = (err, objMD, callback) => {
    // if general error from metadata return error
    if (err) {
        monitoring.promMetrics('DELETE', bucketName, err.code,
            'multiObjectDelete');
        return callback(err);
    }
    if (!objMD) {
        const verCfg = bucket.getVersioningConfiguration();
        // To adhere to AWS behavior, create a delete marker
        // if trying to delete an object that does not exist
        // when versioning has been configured
        if (verCfg && !entry.versionId) {
            log.debug('trying to delete specific version ' +
                ' that does not exist');
            return callback(null, objMD, versionId);
        }
        // otherwise if particular key does not exist, AWS
        // returns success for key so add to successfullyDeleted
        // list and move on
        successfullyDeleted.push({ entry });
        return callback(skipError);
    }
    if (versionId && objMD.location &&
        Array.isArray(objMD.location) && objMD.location[0]) {
        // we need this information for data deletes to AWS
        // eslint-disable-next-line no-param-reassign
        objMD.location[0].deleteVersion = true;
    }
    return callback(null, objMD, versionId);
};
...
(versionId, callback) => {
    if (cache && cache[`${entry.key}${versionId}`]) {
        return handleObjectMetadata(null, cache[`${entry.key}${versionId}`], callback);
    } else {
        return metadataGetObject(bucketName, entry.key, versionId, log, (err, objMD) => handleObjectMetadata(err, objMD, callback));
    }
...
```

Reply:
Ah OK with the `cachedDocument` param, fine for me.

I don't know if it's the best approach but should work, personally I'd rather have abstracted the `metadataGetObjects` implementation itself into a single call to `getObjectsMD` or an async loop of `getObjectMD` calls (with concurrency limit), but if you want to minimize latency you may want to start processing results as soon as they arrive instead if the backend doesn't support `getObjectsMD`, however it's probably not a significant optimization in my sense.

Reply:
c.f. my next comment: I guess this would be much clearer if the `callback` was kept as last param
Having the cache after the callback makes it very hard to read the call...